vm/devices/storage/disk_backend/src/layered.rs

```diff
@@ -0,0 +1,718 @@
+        &self,
+        buffers: &RequestBuffers<'_>,
+        sector: u64,
+    ) -> impl Future<Output = Result<(), DiskError>> + Send;
+}
+
+impl<T: WriteNoOverwrite> WriteNoOverwrite for &T {
+    fn write_no_overwrite(
+        &self,
+        buffers: &RequestBuffers<'_>,
+        sector: u64,
+    ) -> impl Future<Output = Result<(), DiskError>> + Send {
+        (*self).write_no_overwrite(buffers, sector)
+    }
+}
+
+impl WriteNoOverwrite for NoIdet {
+    async fn write_no_overwrite(
+        &self,
+        _buffers: &RequestBuffers<'_>,
+        _sector: u64,
+    ) -> Result<(), DiskError> {
+        unreachable!()
+    }
+}
+
+impl DiskIo for LayeredDisk {
+    fn disk_type(&self) -> &str {
+        "layered"
+    }
+
+    fn sector_count(&self) -> u64 {
+        self.layers[0].backing.sector_count()
+    }
+
+    fn sector_size(&self) -> u32 {
+        1 << self.sector_shift
+    }
+
+    fn disk_id(&self) -> Option<[u8; 16]> {
+        self.disk_id
+    }
+
+    fn physical_sector_size(&self) -> u32 {
+        self.physical_sector_size
+    }
+
+    fn is_fua_respected(&self) -> bool {
+        false
+    }
+
+    fn is_read_only(&self) -> bool {
+        self.is_read_only
+    }
+
+    async fn read_vectored(
+        &self,
+        buffers: &RequestBuffers<'_>,
+        sector: u64,
+    ) -> Result<(), DiskError> {
+        let sector_count = buffers.len() >> self.sector_shift;
+        let mut bitmap = Bitmap::new(sector, sector_count);
+        let mut bits_set = 0;
+        // FUTURE: queue the reads to the layers in parallel.
+        'done: for (i, layer) in self.layers.iter().enumerate() {
+            if bits_set == sector_count {
+                break;
+            }
+            for mut range in bitmap.unset_iter() {
+                let end = if i == 0 {
+                    // The visible sector count of the first layer is unknown,
+                    // since it could change at any time.
+                    range.end_sector()
+                } else {
+                    // Restrict the range to the visible sector count of the
+                    // layer; sectors beyond this are logically zero.
+                    let end = range.end_sector().min(layer.visible_sector_count);
+                    if range.start_sector() == end {
+                        break 'done;
+                    }
+                    end
+                };
+
+                let sectors = end - range.start_sector();
+
+                let buffers = buffers.subrange(
+                    range.start_sector_within_bitmap() << self.sector_shift,
+                    (sectors as usize) << self.sector_shift,
+                );
+
+                layer
+                    .backing
+                    .read(&buffers, range.start_sector(), range.view(sectors))
+                    .await?;
+
+                bits_set += range.set_count();
+
+                // TODO: populate read cache(s). Note that we need to detect
```

Comment:
I have no qualms with this TODO, since I don't actually need read cache functionality for some of the test infra stuff I want to hack on. migrating certain test infra over to blob_disk would be great, but it continues to be non-critical